pkg/rslog/debug.go

```diff
@@ -118,86 +137,73 @@ type DebugLogger interface {
 
 type debugLogger struct {
 	Logger
-	region  ProductRegion
-	enabled bool
+	region ProductRegion
 }
 
-// NewDebugLogger returns a new logger which includes
-// the name of the debug region at every message.
+// NewDebugLogger returns a new logger which includes the name of the product
+// debug region at every message. Debugf and Tracef only occur when the
+// product debug region is enabled.
+//
+// Logging occurs only when enabled for region.
 func NewDebugLogger(region ProductRegion) *debugLogger {
 	lgr := DefaultLogger()
 
+	regionName := RegionName(region)
 	entry := lgr.WithFields(Fields{
-		"region": regionNames[region],
+		"region": regionName,
 	})
 
 	dbglgr := &debugLogger{
-		Logger:  entry,
-		region:  region,
-		enabled: Enabled(region),
+		Logger: entry,
+		region: region,
 	}
 
-	registerLoggerCb(region, dbglgr.enable)
-
 	return dbglgr
 }
 
-func registerLoggerCb(region ProductRegion, cb func(bool)) {
-	debugMutex.Lock()
-	defer debugMutex.Unlock()
-
-	regionCallbacks[region] = append(regionCallbacks[region], cb)
-}
-
-// Enabled returns true if debug logging is enabled for this rslog.
+// Enabled returns true if debug logging is enabled for the associated product
+// debug region.
 func (l *debugLogger) Enabled() bool {
 	return Enabled(l.region)
 }
 
+// Debugf logs a message when debug logging is enabled for the associated
+// product debug region.
 func (l *debugLogger) Debugf(message string, args ...interface{}) {
-	debugMutex.RLock()
-	defer debugMutex.RUnlock()
-
-	if l.enabled {
+	if l.Enabled() {
 		l.Logger.Debugf(message, args...)
 	}
 }
 
+// Tracef logs a message when debug logging is enabled for the associated
+// product debug region.
 func (l *debugLogger) Tracef(message string, args ...interface{}) {
-	debugMutex.RLock()
-	defer debugMutex.RUnlock()
-
-	if l.enabled {
+	if l.Enabled() {
 		l.Logger.Tracef(message, args...)
 	}
 }
 
-// Set fields to be logged
+// WithFields returns a new logger having additional fields. The returned
+// logger is associated with the same product debug region.
 func (l *debugLogger) WithFields(fields Fields) DebugLogger {
 	newLgr := l.Logger.WithFields(fields)
 	dbglgr := &debugLogger{
-		Logger:  newLgr,
-		region:  l.region,
-		enabled: l.enabled,
+		Logger: newLgr,
+		region: l.region,
 	}
-	registerLoggerCb(l.region, dbglgr.enable)
 	return dbglgr
 }
 
-// WithSubRegion returns a debug logger with further specificity
-// via sub_region key:value. E.g "region": "LDAP", "sub_region": "membership scanner"
+// WithSubRegion returns a debug logger having an additional "sub_region"
+// field. The returned logger is associated with the same product debug
+// region.
+//
+// Equivalent to `debugLogger.WithField("sub_region", subregion)`
 func (l *debugLogger) WithSubRegion(subregion string) DebugLogger {
 	newLgr := l.Logger.WithField("sub_region", subregion)
```

Comment:
This isn't really relevant for this PR but I don't understand why `WithSubRegion` exists. Why not just call `WithFields()` like we do in the tests down below?

```go
                gr.WithFields(Fields{
			"region":     "alfa",
			"sub_region": "drinks",
		})
```

Reply:
I do not want to alter the public API.

Reply:
The API exists because folks wanted a well-known field name.